jupyterlab_gcedetails/src/service/notebook_instance_service_layer.ts

```diff
@@ -0,0 +1,453 @@
+    return err.result.error.code === GOOGLE_API_UNAUTHENITCATED_CODE;
+  }
+
+  return false;
+}
+
+/**
+ * Wraps around the GCP AI Platform Notebooks API to manage notebook resources in Google Cloud.
+ * https://cloud.google.com/ai-platform/notebooks/docs/reference/rest
+ */
+export class NotebookInstanceServiceLayer {
+  private projectIdPromise?: Promise<string>;
+  private instanceNamePromise?: Promise<string>;
+  private locationIdPromise?: Promise<string>;
+  private metadataPromise?: Promise<InstanceMetadata>;
+
+  constructor(
+    private _transportService: TransportService,
+    private _authToken: string,
+    projectId?: string,
+    instanceName?: string,
+    locationId?: string
+  ) {
+    if (projectId) {
+      this.projectIdPromise = Promise.resolve(projectId);
+    }
+    if (instanceName) {
+      this.instanceNamePromise = Promise.resolve(instanceName);
+    }
+    if (locationId) {
+      this.locationIdPromise = Promise.resolve(locationId);
+    }
+  }
+
+  get projectId() {
+    return this._getProject();
+  }
+
+  set projectId(projectIdPromise: string | Promise<string>) {
+    if (typeof projectIdPromise === 'string') {
+      this.projectIdPromise = Promise.resolve(projectIdPromise);
+    } else {
+      this.projectIdPromise = projectIdPromise;
+    }
+  }
+
+  get instanceName() {
+    return this._getInstance();
+  }
+
+  set instanceName(instanceNamePromise: string | Promise<string>) {
+    if (typeof instanceNamePromise === 'string') {
+      this.instanceNamePromise = Promise.resolve(instanceNamePromise);
+    } else {
+      this.instanceNamePromise = instanceNamePromise;
+    }
+  }
+
+  get locationId() {
+    return this._getLocation();
+  }
+
+  set locationId(locationIdPromise: string | Promise<string>) {
+    if (typeof locationIdPromise === 'string') {
+      this.locationIdPromise = Promise.resolve(locationIdPromise);
+    } else {
+      this.locationIdPromise = locationIdPromise;
+    }
+  }
+
+  set transportService(transportService: TransportService) {
+    this._transportService = transportService;
+  }
+
+  set authToken(authToken: string) {
+    this._authToken = authToken;
+  }
+
+  /**
+   * Retrieves all necessary details about the user's project to allow the
+   * extension to know whether or not Notebooks can be scheduled.
+   */
+  async getState(): Promise<State> {
+    try {
+      const [projectId, instanceName, locationId] = await Promise.all([
+        this.projectId,
+        this.instanceName,
+        this.locationId,
+      ]);
+
+      const state: State = {
+        serviceEnabled: false,
+        projectId,
+        instanceName,
+        locationId,
+      };
+
+      return state;
+    } catch (err) {
+      console.error('Unable to determine project status', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Enables the service specified returning a Promise for the
+   * operation.
+   */
+  async enableService(serviceName: string): Promise<Operation> {
```

Comment:
As mentioned above, this is not necessary given that a user of a CAIP Notebook will have to already have enabled the API.